File: services/ingestion/connectors/filesystem.py
"""Filesystem connector for local and S3-compatible storage.

This module provides a connector for ingesting documents from local
filesystem and S3-compatible storage (including MinIO).
"""

import mimetypes
from collections.abc import AsyncIterator
from datetime import UTC, datetime
from io import BytesIO
from pathlib import Path
from typing import Literal

import aioboto3
import aiofiles
import aiofiles.os
from pydantic import BaseModel, Field

from .base import (
    BaseConnector,
    DocumentMetadata,
    RawDocument,
)

# Try to import python-magic for better MIME type detection
try:
    import magic

    HAS_MAGIC = True
except ImportError:
    HAS_MAGIC = False

# Register additional MIME types not in the standard library
mimetypes.add_type("text/markdown", ".md")
mimetypes.add_type("text/markdown", ".markdown")


class FilesystemConnectorConfig(BaseModel):
    """Configuration for the filesystem connector.

    Supports both local filesystem and S3-compatible storage like MinIO.
    """

    base_path: str = Field(
        ...,
        description="Base path for local storage or bucket name for S3",
    )
    storage_type: Literal["local", "s3"] = Field(
        default="local",
        description="Type of storage backend",
    )
    s3_endpoint: str | None = Field(
        default=None,
        description="S3/MinIO endpoint URL (e.g., 'http://localhost:9000')",
    )
    s3_access_key: str | None = Field(
        default=None,
        description="S3 access key ID",
    )
    s3_secret_key: str | None = Field(
        default=None,
        description="S3 secret access key",
    )
    s3_bucket: str | None = Field(
        default=None,
        description="S3 bucket name (overrides base_path for S3)",
    )
    s3_region: str | None = Field(
        default="us-east-1",
        description="S3 region name",
    )
    recursive: bool = Field(
        default=True,
        description="Whether to scan directories recursively",
    )
    file_extensions: list[str] | None = Field(
        default=None,
        description="List of file extensions to include (e.g., ['.pdf', '.docx'])",
    )

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "base_path": "/data/documents",
                    "storage_type": "local",
                    "recursive": True,
                    "file_extensions": [".pdf", ".docx", ".txt"],
                },
                {
                    "base_path": "documents",
                    "storage_type": "s3",
                    "s3_endpoint": "http://localhost:9000",
                    "s3_access_key": "minioadmin",
                    "s3_secret_key": "minioadmin123",
                    "s3_bucket": "documents",
                    "s3_region": "us-east-1",
                },
            ],
        },
    }


class FilesystemConnector(BaseConnector):
    """Connector for local filesystem and S3-compatible storage.

    Supports reading documents from:
    - Local filesystem with async I/O (via aiofiles)
    - S3-compatible storage including MinIO (via aioboto3)

    Example:
        ```python
        # Local filesystem
        config = FilesystemConnectorConfig(
            base_path="/data/documents",
            storage_type="local",
            file_extensions=[".pdf", ".txt"]
        )
        async with FilesystemConnector(config) as connector:
            async for doc in connector.stream_documents():
                print(f"Loaded: {doc.metadata.filename}")

        # S3/MinIO
        config = FilesystemConnectorConfig(
            base_path="my-bucket",
            storage_type="s3",
            s3_endpoint="http://localhost:9000",
            s3_access_key="minioadmin",
            s3_secret_key="secret"
        )
        async with FilesystemConnector(config) as connector:
            async for doc in connector.stream_documents("prefix/"):
                process(doc)
        ```
    """

    def __init__(self, config: FilesystemConnectorConfig):
        """Initialize the filesystem connector.

        Args:
            config: Configuration for the connector.
        """
        self.config = config
        self._session: aioboto3.Session | None = None
        self._s3_client = None
        self._connected = False

    async def connect(self) -> None:
        """Establish connection to the storage backend.

        For local storage, this validates the base path exists.
        For S3, this initializes the aioboto3 session and client.

        Raises:
            ConnectionError: If the storage backend is not accessible.
        """
        if self.config.storage_type == "local":
            base = Path(self.config.base_path)
            if not base.exists():
                raise ConnectionError(
                    f"Base path does not exist: {self.config.base_path}",
                )
            self._connected = True
        else:
            # S3 storage
            self._session = aioboto3.Session()
            try:
                # Test connection by listing bucket
                async with self._session.client(
                    "s3",
                    endpoint_url=self.config.s3_endpoint,
                    aws_access_key_id=self.config.s3_access_key,
                    aws_secret_access_key=self.config.s3_secret_key,
                    region_name=self.config.s3_region,
                ) as client:
                    bucket = self.config.s3_bucket or self.config.base_path
                    await client.head_bucket(Bucket=bucket)
            except Exception as e:
                raise ConnectionError(f"Failed to connect to S3: {e}") from e
            self._connected = True

    async def disconnect(self) -> None:
        """Close connection to the storage backend.

        For S3, this cleans up the session.
        """
        self._session = None
        self._s3_client = None
        self._connected = False

    def _should_include_file(self, filename: str) -> bool:
        """Check if a file should be included based on extension filter.

        Args:
            filename: Name of the file to check.

        Returns:
            True if the file should be included, False otherwise.
        """
        if self.config.file_extensions is None:
            return True
        ext = Path(filename).suffix.lower()
        return ext in [e.lower() for e in self.config.file_extensions]

    def _detect_mime_type(self, content: bytes, filename: str | None = None) -> str:
        """Detect the MIME type of file content.

        Uses python-magic if available, falls back to mimetypes.

        Args:
            content: File content as bytes.
            filename: Optional filename for extension-based detection.

        Returns:
            MIME type string.
        """
        if HAS_MAGIC:
            try:
                return magic.from_buffer(content, mime=True)
            except Exception:  # noqa: S110
                pass

        if filename:
            mime_type, _ = mimetypes.guess_type(filename)
            if mime_type:
                return mime_type

        return "application/octet-stream"

    async def _list_local_files(
        self,
        path: str | None = None,
    ) -> AsyncIterator[DocumentMetadata]:
        """List files from local filesystem.

        Args:
            path: Optional subdirectory to list.

        Yields:
            DocumentMetadata for each file found.
        """
        base = Path(self.config.base_path)
        if path:
            base = base / path

        if not base.exists():
            return

        if base.is_file():
            if self._should_include_file(base.name):
                stat = await aiofiles.os.stat(base)
                yield DocumentMetadata(
                    source_id=str(base.relative_to(self.config.base_path)),
                    source_type="local",
                    filename=base.name,
                    created_at=datetime.fromtimestamp(stat.st_ctime, tz=UTC),
                    modified_at=datetime.fromtimestamp(stat.st_mtime, tz=UTC),
                    size_bytes=stat.st_size,
                    extra={"absolute_path": str(base.absolute())},
                )
            return

        # Directory traversal
        if self.config.recursive:
            for item in base.rglob("*"):
                if item.is_file() and self._should_include_file(item.name):
                    stat = await aiofiles.os.stat(item)
                    yield DocumentMetadata(
                        source_id=str(item.relative_to(self.config.base_path)),
                        source_type="local",
                        filename=item.name,
                        created_at=datetime.fromtimestamp(stat.st_ctime, tz=UTC),
                        modified_at=datetime.fromtimestamp(stat.st_mtime, tz=UTC),
                        size_bytes=stat.st_size,
                        extra={"absolute_path": str(item.absolute())},
                    )
        else:
            for item in base.iterdir():
                if item.is_file() and self._should_include_file(item.name):
                    stat = await aiofiles.os.stat(item)
                    yield DocumentMetadata(
                        source_id=str(item.relative_to(self.config.base_path)),
                        source_type="local",
                        filename=item.name,
                        created_at=datetime.fromtimestamp(stat.st_ctime, tz=UTC),
                        modified_at=datetime.fromtimestamp(stat.st_mtime, tz=UTC),
                        size_bytes=stat.st_size,
                        extra={"absolute_path": str(item.absolute())},
                    )

    async def _list_s3_files(
        self,
        path: str | None = None,
    ) -> AsyncIterator[DocumentMetadata]:
        """List files from S3-compatible storage.

        Args:
            path: Optional prefix to filter objects. If not provided,
                  uses base_path from config as the default prefix.

        Yields:
            DocumentMetadata for each object found.
        """
        bucket = self.config.s3_bucket or self.config.base_path
        # Use base_path as default prefix when path not specified
        prefix = path if path is not None else self.config.base_path

        async with self._session.client(
            "s3",
            endpoint_url=self.config.s3_endpoint,
            aws_access_key_id=self.config.s3_access_key,
            aws_secret_access_key=self.config.s3_secret_key,
            region_name=self.config.s3_region,
        ) as client:
            paginator = client.get_paginator("list_objects_v2")
            async for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
                for obj in page.get("Contents", []):
                    key = obj["Key"]
                    filename = key.split("/")[-1] if "/" in key else key

                    # Skip "directory" markers
                    if key.endswith("/"):
                        continue

                    if not self._should_include_file(filename):
                        continue

                    yield DocumentMetadata(
                        source_id=key,
                        source_type="s3",
                        filename=filename,
                        modified_at=obj.get("LastModified"),
                        size_bytes=obj.get("Size"),
                        extra={
                            "bucket": bucket,
                            "etag": obj.get("ETag", "").strip('"'),
                            "storage_class": obj.get("StorageClass"),
                        },
                    )

    async def list_documents(
        self,
        path: str | None = None,
    ) -> AsyncIterator[DocumentMetadata]:
        """List available documents at the given path.

        Args:
            path: For local storage, a subdirectory path.
                  For S3, an object key prefix.

        Yields:
            DocumentMetadata for each document found.

        Raises:
            ConnectionError: If not connected to the storage backend.
        """
        if not self._connected:
            raise ConnectionError("Connector is not connected. Call connect() first.")

        if self.config.storage_type == "local":
            async for meta in self._list_local_files(path):
                yield meta
        else:
            async for meta in self._list_s3_files(path):
                yield meta

    async def _fetch_local_document(self, source_id: str) -> RawDocument:
        """Fetch a document from local filesystem.

        Args:
            source_id: Relative path to the file.

        Returns:
            RawDocument with content and metadata.

        Raises:
            FileNotFoundError: If the file does not exist.
        """
        file_path = Path(self.config.base_path) / source_id

        if not file_path.exists():
            raise FileNotFoundError(f"File not found: {source_id}")

        if not file_path.is_file():
            raise FileNotFoundError(f"Not a file: {source_id}")

        async with aiofiles.open(file_path, "rb") as f:
            content = await f.read()

        stat = await aiofiles.os.stat(file_path)
        mime_type = self._detect_mime_type(content, file_path.name)

        metadata = DocumentMetadata(
            source_id=source_id,
            source_type="local",
            filename=file_path.name,
            mime_type=mime_type,
            created_at=datetime.fromtimestamp(stat.st_ctime, tz=UTC),
            modified_at=datetime.fromtimestamp(stat.st_mtime, tz=UTC),
            size_bytes=len(content),
            extra={"absolute_path": str(file_path.absolute())},
        )

        return RawDocument(content=content, metadata=metadata)

    async def _fetch_s3_document(self, source_id: str) -> RawDocument:
        """Fetch a document from S3-compatible storage.

        Args:
            source_id: Object key in the bucket.

        Returns:
            RawDocument with content and metadata.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        bucket = self.config.s3_bucket or self.config.base_path

        async with self._session.client(
            "s3",
            endpoint_url=self.config.s3_endpoint,
            aws_access_key_id=self.config.s3_access_key,
            aws_secret_access_key=self.config.s3_secret_key,
            region_name=self.config.s3_region,
        ) as client:
            try:
                # Get object content
                buffer = BytesIO()
                await client.download_fileobj(
                    Bucket=bucket,
                    Key=source_id,
                    Fileobj=buffer,
                )
                content = buffer.getvalue()

                # Get object metadata
                head = await client.head_object(Bucket=bucket, Key=source_id)

            except client.exceptions.NoSuchKey:
                raise FileNotFoundError(f"Object not found: {source_id}") from None
            except Exception as e:
                if "NoSuchKey" in str(e) or "404" in str(e):
                    raise FileNotFoundError(f"Object not found: {source_id}") from e
                raise

        # Prefer original filename from S3 metadata if available
        s3_metadata = head.get("Metadata", {})
        original_filename = s3_metadata.get("original-filename")
        if original_filename:
            filename = original_filename
        else:
            # Fall back to extracting from S3 key
            key_filename = source_id.split("/")[-1] if "/" in source_id else source_id
            # Strip timestamp prefix if present (format: {timestamp}-{filename})
            if "-" in key_filename and key_filename.split("-")[0].isdigit():
                filename = key_filename.split("-", 1)[1]
            else:
                filename = key_filename
        mime_type = head.get("ContentType") or self._detect_mime_type(content, filename)

        metadata = DocumentMetadata(
            source_id=source_id,
            source_type="s3",
            filename=filename,
            mime_type=mime_type,
            modified_at=head.get("LastModified"),
            size_bytes=len(content),
            extra={
                "bucket": bucket,
                "etag": head.get("ETag", "").strip('"'),
                "content_encoding": head.get("ContentEncoding"),
                "version_id": head.get("VersionId"),
                "original_filename": original_filename,
                "s3_key": source_id,
            },
        )

        return RawDocument(content=content, metadata=metadata)

    async def fetch_document(self, source_id: str) -> RawDocument:
        """Fetch a single document by its source ID.

        Args:
            source_id: For local storage, relative file path.
                       For S3, object key.

        Returns:
            RawDocument containing the document content and metadata.

        Raises:
            ConnectionError: If not connected to the storage backend.
            FileNotFoundError: If the document does not exist.
        """
        if not self._connected:
            raise ConnectionError("Connector is not connected. Call connect() first.")

        if self.config.storage_type == "local":
            return await self._fetch_local_document(source_id)
        return await self._fetch_s3_document(source_id)

    async def stream_documents(
        self,
        path: str | None = None,
    ) -> AsyncIterator[RawDocument]:
        """Stream all documents from the given path.

        Combines listing and fetching, yielding complete documents
        one at a time. This is memory-efficient for processing
        large numbers of documents.

        Args:
            path: For local storage, a subdirectory path.
                  For S3, an object key prefix.

        Yields:
            RawDocument for each document at the path.

        Raises:
            ConnectionError: If not connected to the storage backend.
        """
        if not self._connected:
            raise ConnectionError("Connector is not connected. Call connect() first.")

        async for metadata in self.list_documents(path):
            try:
                doc = await self.fetch_document(metadata.source_id)
                yield doc
            except FileNotFoundError:
                # File may have been deleted between listing and fetching
                continue
            except Exception as e:
                # Log error but continue with other documents
                # In production, this should use proper logging
                print(f"Error fetching {metadata.source_id}: {e}")
                continue

File: crates/rag-ingestion/src/connectors/filesystem.rs
//! Filesystem connector for local documents.

use async_trait::async_trait;
use bytes::Bytes;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use tokio::fs;
use tracing::{debug, instrument};

use super::base::{Connector, DocumentMetadata, RawDocument, StorageBackend};
use crate::error::{Error, Result};

/// Configuration for the filesystem connector.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesystemConfig {
    /// Base path to scan for documents.
    pub base_path: PathBuf,
    /// Whether to scan subdirectories recursively.
    #[serde(default = "default_recursive")]
    pub recursive: bool,
    /// File extensions to include (e.g., [".pdf", ".md"]).
    /// If None, all files are included.
    pub file_extensions: Option<Vec<String>>,
}

fn default_recursive() -> bool {
    true
}

impl FilesystemConfig {
    /// Create a new configuration for the given path.
    pub fn new(base_path: impl Into<PathBuf>) -> Self {
        Self {
            base_path: base_path.into(),
            recursive: true,
            file_extensions: None,
        }
    }

    /// Set whether to scan recursively.
    #[must_use]
    pub fn with_recursive(mut self, recursive: bool) -> Self {
        self.recursive = recursive;
        self
    }

    /// Set file extensions to filter.
    #[must_use]
    pub fn with_extensions(mut self, extensions: Vec<String>) -> Self {
        self.file_extensions = Some(extensions);
        self
    }
}

/// Connector for local filesystem documents.
pub struct FilesystemConnector {
    config: FilesystemConfig,
    connected: bool,
}

impl FilesystemConnector {
    /// Create a new filesystem connector.
    pub fn new(config: FilesystemConfig) -> Self {
        Self {
            config,
            connected: false,
        }
    }

    /// Check if a file should be included based on extension filter.
    fn should_include(&self, path: &Path) -> bool {
        match &self.config.file_extensions {
            None => true,
            Some(extensions) => path
                .extension()
                .and_then(|ext| ext.to_str())
                .map(|ext| {
                    let ext_with_dot = format!(".{}", ext.to_lowercase());
                    extensions.iter().any(|e| e.to_lowercase() == ext_with_dot)
                })
                .unwrap_or(false),
        }
    }

    /// Recursively scan a directory for files.
    async fn scan_directory(&self, path: &Path) -> Result<Vec<PathBuf>> {
        let mut files = Vec::new();
        let mut dirs_to_scan = vec![path.to_path_buf()];

        while let Some(dir) = dirs_to_scan.pop() {
            let mut entries = fs::read_dir(&dir).await.map_err(|e| {
                Error::Connector(format!("Failed to read directory {}: {}", dir.display(), e))
            })?;

            while let Some(entry) = entries
                .next_entry()
                .await
                .map_err(|e| Error::Connector(format!("Failed to read directory entry: {e}")))?
            {
                let entry_path = entry.path();
                let file_type = entry
                    .file_type()
                    .await
                    .map_err(|e| Error::Connector(format!("Failed to get file type: {e}")))?;

                if file_type.is_dir() {
                    if self.config.recursive {
                        dirs_to_scan.push(entry_path);
                    }
                } else if file_type.is_file() && self.should_include(&entry_path) {
                    files.push(entry_path);
                }
            }
        }

        Ok(files)
    }

    /// Get metadata for a file.
    async fn get_file_metadata(&self, path: &Path) -> Result<DocumentMetadata> {
        let metadata = fs::metadata(path).await.map_err(|e| {
            Error::Connector(format!(
                "Failed to get metadata for {}: {}",
                path.display(),
                e
            ))
        })?;

        let source_id = path
            .strip_prefix(&self.config.base_path)
            .unwrap_or(path)
            .to_string_lossy()
            .to_string();

        let filename = path
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default();

        let mime_type = mime_guess::from_path(path).first().map(|m| m.to_string());

        let size_bytes = metadata.len();

        let modified_at = metadata.modified().ok().map(DateTime::<Utc>::from);

        let created_at = metadata.created().ok().map(DateTime::<Utc>::from);

        Ok(
            DocumentMetadata::new(source_id, StorageBackend::Local, filename)
                .with_size(size_bytes)
                .with_timestamps(created_at, modified_at)
                .with_mime_type(
                    mime_type.unwrap_or_else(|| "application/octet-stream".to_string()),
                ),
        )
    }
}

#[async_trait]
impl Connector for FilesystemConnector {
    #[instrument(skip(self))]
    async fn connect(&mut self) -> Result<()> {
        // Verify the base path exists and is a directory
        let metadata = fs::metadata(&self.config.base_path).await.map_err(|e| {
            Error::Connector(format!(
                "Base path {} does not exist or is not accessible: {}",
                self.config.base_path.display(),
                e
            ))
        })?;

        if !metadata.is_dir() {
            return Err(Error::Connector(format!(
                "Base path {} is not a directory",
                self.config.base_path.display()
            )));
        }

        self.connected = true;
        debug!(base_path = %self.config.base_path.display(), "Connected to filesystem");
        Ok(())
    }

    async fn disconnect(&mut self) -> Result<()> {
        self.connected = false;
        debug!("Disconnected from filesystem");
        Ok(())
    }

    #[instrument(skip(self))]
    async fn list_documents(&self, path: Option<&str>) -> Result<Vec<DocumentMetadata>> {
        if !self.connected {
            return Err(Error::Connector("Not connected".to_string()));
        }

        let scan_path = match path {
            Some(p) => self.config.base_path.join(p),
            None => self.config.base_path.clone(),
        };

        let files = self.scan_directory(&scan_path).await?;
        let mut documents = Vec::with_capacity(files.len());

        for file_path in files {
            match self.get_file_metadata(&file_path).await {
                Ok(meta) => documents.push(meta),
                Err(e) => {
                    debug!(path = %file_path.display(), error = %e, "Skipping file");
                }
            }
        }

        Ok(documents)
    }

    #[instrument(skip(self))]
    async fn fetch_document(&self, source_id: &str) -> Result<RawDocument> {
        if !self.connected {
            return Err(Error::Connector("Not connected".to_string()));
        }

        let file_path = self.config.base_path.join(source_id);

        if !file_path.exists() {
            return Err(Error::NotFound(format!("Document not found: {source_id}")));
        }

        let content = fs::read(&file_path).await.map_err(|e| {
            Error::Connector(format!("Failed to read {}: {}", file_path.display(), e))
        })?;

        let metadata = self.get_file_metadata(&file_path).await?;

        Ok(RawDocument::new(Bytes::from(content), metadata))
    }

    fn is_connected(&self) -> bool {
        self.connected
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;
    use tokio::io::AsyncWriteExt;

    async fn setup_test_dir() -> TempDir {
        let dir = TempDir::new().unwrap();

        // Create test files
        let file1 = dir.path().join("test.txt");
        let mut f = tokio::fs::File::create(&file1).await.unwrap();
        f.write_all(b"Hello, World!").await.unwrap();

        let file2 = dir.path().join("test.md");
        let mut f = tokio::fs::File::create(&file2).await.unwrap();
        f.write_all(b"# Markdown").await.unwrap();

        // Create subdirectory with file
        let subdir = dir.path().join("subdir");
        tokio::fs::create_dir(&subdir).await.unwrap();
        let file3 = subdir.join("nested.txt");
        let mut f = tokio::fs::File::create(&file3).await.unwrap();
        f.write_all(b"Nested content").await.unwrap();

        dir
    }

    #[tokio::test]
    async fn test_connect_success() {
        let dir = setup_test_dir().await;
        let config = FilesystemConfig::new(dir.path());
        let mut connector = FilesystemConnector::new(config);

        assert!(!connector.is_connected());
        connector.connect().await.unwrap();
        assert!(connector.is_connected());
    }

    #[tokio::test]
    async fn test_connect_nonexistent_path() {
        let config = FilesystemConfig::new("/nonexistent/path");
        let mut connector = FilesystemConnector::new(config);

        let result = connector.connect().await;
        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_list_documents_all() {
        let dir = setup_test_dir().await;
        let config = FilesystemConfig::new(dir.path());
        let mut connector = FilesystemConnector::new(config);
        connector.connect().await.unwrap();

        let docs = connector.list_documents(None).await.unwrap();
        assert_eq!(docs.len(), 3); // test.txt, test.md, subdir/nested.txt
    }

    #[tokio::test]
    async fn test_list_documents_filtered() {
        let dir = setup_test_dir().await;
        let config = FilesystemConfig::new(dir.path()).with_extensions(vec![".txt".to_string()]);
        let mut connector = FilesystemConnector::new(config);
        connector.connect().await.unwrap();

        let docs = connector.list_documents(None).await.unwrap();
        assert_eq!(docs.len(), 2); // test.txt, subdir/nested.txt
        assert!(docs.iter().all(|d| d.filename.ends_with(".txt")));
    }

    #[tokio::test]
    async fn test_list_documents_non_recursive() {
        let dir = setup_test_dir().await;
        let config = FilesystemConfig::new(dir.path()).with_recursive(false);
        let mut connector = FilesystemConnector::new(config);
        connector.connect().await.unwrap();

        let docs = connector.list_documents(None).await.unwrap();
        assert_eq!(docs.len(), 2); // test.txt, test.md (no nested)
    }

    #[tokio::test]
    async fn test_fetch_document() {
        let dir = setup_test_dir().await;
        let config = FilesystemConfig::new(dir.path());
        let mut connector = FilesystemConnector::new(config);
        connector.connect().await.unwrap();

        let doc = connector.fetch_document("test.txt").await.unwrap();
        assert_eq!(doc.content_as_str().unwrap(), "Hello, World!");
        assert_eq!(doc.metadata.filename, "test.txt");
        assert_eq!(doc.metadata.source_type, StorageBackend::Local);
    }

    #[tokio::test]
    async fn test_fetch_document_not_found() {
        let dir = setup_test_dir().await;
        let config = FilesystemConfig::new(dir.path());
        let mut connector = FilesystemConnector::new(config);
        connector.connect().await.unwrap();

        let result = connector.fetch_document("nonexistent.txt").await;
        assert!(matches!(result, Err(Error::NotFound(_))));
    }

    #[tokio::test]
    async fn test_fetch_without_connect() {
        let dir = setup_test_dir().await;
        let config = FilesystemConfig::new(dir.path());
        let connector = FilesystemConnector::new(config);

        let result = connector.fetch_document("test.txt").await;
        assert!(result.is_err());
    }
}
